src/main/scala/sfs.scala

```diff
@@ -0,0 +1,46 @@
+package suffuse
+package api
+
+trait Filesystem {
+  /** A Path is a serialization of the steps one must take from
+   *  the root to a particular node in the tree. A Name is a single
+   *  directed edge of the graph.
+   */
+  type Path
+  type Name
+
+  /** A key of some kind, only meaningful within this filesystem,
+   *  which uniquely pinpoints the entity to which a path resolves.
+   *  It is decoupled from the name.
+   */
+  type Key
+
+  /** Some means of performing I/O on a virtualized file.
+   */
+  type FileIO
+
+  /** There are a number of ways we could arrange the "primitives"
+   *  here. The most important thing is to retain the decoupling at
+   *  abstraction boundaries so that one does not wind up computing
+   *  things too eagerly, nor caching data for too long.
+   */
+  def resolve(path: Path): Key
+  def metadata(key: Key): Metadata
+  def lookup(key: Key): Data
```

Comment:
...and already we have an example of why I don't want to call the other Key Key.